### PR TITLE
[SPARK-11633] [SQL] LogicalRDD throws TreeNode Exception : Failed to Copy Node

### PR DESCRIPTION
**Pull request by gatorsmile**
When handling self joins, the implementation did not consider the case insensitivity of HiveContext. It could cause an exception as shown in the JIRA: 

```
TreeNodeException: Failed to copy node. 
```

The fix is low risk. It avoids unnecessary attribute replacement. It should not affect the existing behavior of self joins. Also added the test case to cover this case.

### CODE REVIEW

**Comment by cloud-fan** [comment]
it looks to me that it's a bug of [`LogicalRDD`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala#L70-L73), when we use curry constructor, we should overwrite `otherCopyArgs` that defined in `TreeNode`, like what we do in [`Alias`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala#L157-L159)

**Comment by marmbrus** [comment]
@cloud-fan s suggestion sounds correct to me.

**Comment by gatorsmile** [comment]
@cloud-fan @marmbrus Will follow your suggestions to update the fix. Thanks!

**Comment by gatorsmile** [comment]
@marmbrus @cloud-fan Based on your comments, I did the change. Please review the new change. 

I also tried the fix after excluding the change in `attributeRewrites`. The newly introduced test case still works fine. That means this should be the root cause. 

The fix still keeps the extra filter in `attributeRewrites`. I think it can avoid extra compare and replacement   in the subsequent transformation. Please let me know if you want to remove the filter. 

Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1479,4 +1481,18 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
         |FROM (SELECT '{"f1": "value1", "f2": 12}' json, 'hello' as str) test
       """.stripMargin), Row("value1", "12", 3.14, "hello"))
   }
+
+  test ("SPARK-11633: HiveContext throws TreeNode Exception : Failed to Copy Node") {
+    val rdd1 = sparkContext.parallelize(Seq( Individual(1, 3), Individual(2, 1)))
+    val df = hiveContext.createDataFrame(rdd1)
+    df.registerTempTable("foo")
```
Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Nit, most tests are now using this more concise style.  You can then avoid adding another class. 
```scala
val df = Seq((1, 3), (2, 1)).toDF("f1", "f2")
df.registerTempTable("foo")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -425,7 +425,8 @@ class Analyzer(
              */
             j
           case Some((oldRelation, newRelation)) =>
-            val attributeRewrites = AttributeMap(oldRelation.output.zip(newRelation.output))
+            val attributeRewrites =
+              AttributeMap(oldRelation.output.zip(newRelation.output).filter(x => x._1 != x._2))
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Why this change?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Hi, @marmbrus 

I also noticed your way is more concise, but it will not create LogicalRDD. Instead, what it created is LocalRelation. That is why I did that in the test case. 

Let me post both logical plans for your convenience.  

```
== Analyzed Logical Plan ==
F1: int
Project [F1#0]
 Join Inner, Some((F2#2 = F2#9))
  Subquery a
   Subquery foo2
    Project [f1#0,F2#1 AS F2#2]
     Subquery foo
      LogicalRDD [F1#0,F2#1], MapPartitionsRDD[1] at apply at Transformer.scala:22
  Subquery b
   Subquery foo3
    Project [F1#12,F2#13 AS F2#9]
     Subquery foo
      LogicalRDD [F1#12,F2#13], MapPartitionsRDD[1] at apply at Transformer.scala:22
```

```
== Analyzed Logical Plan ==
F1: int
Project [F1#2]
 Join Inner, Some((F2#4 = F2#11))
  Subquery a
   Subquery foo2
    Project [f1#2,F2#3 AS F2#4]
     Subquery foo
      Project [_1#0 AS F1#2,_2#1 AS F2#3]
       LocalRelation [_1#0,_2#1], [[1,3],[2,1]]
  Subquery b
   Subquery foo3
    Project [F1#14,F2#15 AS F2#11]
     Subquery foo
      Project [_1#0 AS F1#14,_2#1 AS F2#15]
       LocalRelation [_1#0,_2#1], [[1,3],[2,1]]
```

Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Oh, makes sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: It can avoid extra compare and replacement in the subsequent transformation. For example, in the test case I added, this change can avoid making a copy for the underlying `LogicalRDD`. 

That is my original fix. If you think this change is not needed, I can easily remove it.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I guess I would rather remove it.  In general, we rely on tree node to catch when transformations are no-ops in order to keep the code simple.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Maybe add a comment?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Agree. Will do.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Sure. Will do.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 3]
```diff
@@ -1479,4 +1481,20 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
         |FROM (SELECT '{"f1": "value1", "f2": 12}' json, 'hello' as str) test
       """.stripMargin), Row("value1", "12", 3.14, "hello"))
   }
+
+  // This test case is to verify a bug when making a new instance of LogicalRDD.
+  test ("SPARK-11633: HiveContext throws TreeNode Exception : Failed to Copy Node") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: some suggestion about the test:

* do not put tests in hive module if it's not hive related. Actually we can reproduce this bug by setting the case sensitive config to false and put the test in `DataFrameSuite`.
* do not create new classes unless we have to. For this case we can create RDD[Row] and specify the schema for it.
* clean up temp table at the end of test, or only use the `DataFrame` API

here is an example and hope you can follow it in the future :)
```
withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
  val rdd = sparkContext.makeRDD(Seq(Row(1, 3), Row(2, 1)))
  val df = sqlContext.createDataFrame(
    rdd,
    new StructType().add("f1", IntegerType).add("f2", IntegerType),
    needsConversion = false).select($"F1", $"f2".as("f2"))
  val df1 = df.as("a")
  val df2 = df.as("b")
  checkAnswer(df1.join(df2, $"a.f2" === $"b.f2"), Row(2) :: Row(1) :: Nil)
}
```

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Thank you for sharing this! I will keep it in mind and follow this in the future. 

BTW, if you think it is necessary, I still can change it now. Thanks!

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: please update it, I think the example I gave can reproduce this bug, thanks!

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: I did the try. Yeah, it can reproduce it! I just need to change the expected result. Thank you for your help!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -1997,4 +1997,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     sqlContext.setConf("spark.sql.subexpressionElimination.enabled", "true")
     verifyCallCount(df.selectExpr("testUdf(a)", "testUdf(a)"), Row(1, 1), 1)
   }
+
+  // This test case is to verify a bug when making a new instance of LogicalRDD.
+  test("SPARK-11633: LogicalRDD throws TreeNode Exception : Failed to Copy Node") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: maybe put this in `DataFrameSuite`? we didn't run SQL query in this test.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: : ) True. Your test case does not use SQL.

**Comment by cloud-fan** [comment]
LGTM

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.6
